app/models/interface.rb

```diff
@@ -0,0 +1,30 @@
+class Interface < ApplicationRecord
+  enum interface_type: { discord_guild: 0, web: 1, api: 2 }
+
+  before_save :ensure_single_instance, if: -> { api? || web? }
+
+  has_many :messages, dependent: :destroy
+
+  validates :interface_type, presence: true
+  validates :external_id, presence: true, if: :requires_external_id?
+  validates :external_id, uniqueness: { scope: :interface_type }, if: :requires_external_id?
+
+  private
+
+  def ensure_single_instance
+    existing_interface = Interface.find_by(interface_type:)
+
+    return unless existing_interface
+
+    self.external_id = existing_interface.external_id
+    self.created_at = existing_interface.created_at
+    self.id = existing_interface.id
+
+    # mark it as not new
+    @new_record = false
+  end
```

Comment:
let's avoid using callbacks, please. A *custom validation* is a better fit here.

```rb
validate :unique_source

private

def unique_source
  return unless [:web, :api].include?(source)

  if self.class.where(source:).where.not(id:).present?
    errors.add(:source, "internal sources must be unique"
  end
end
```

Reply:
Done!